homeassistant/components/yolink/__init__.py

```diff
@@ -55,85 +55,92 @@
-                        return
-                    resolved_state = data.data
-                    if resolved_state is None:
-                        return
-                    self.data[message[0]] = resolved_state
-                    self.async_set_updated_data(self.data)
-
                 await self._mqtt_client.init_home_connection(
-                    home_info.data["id"], on_message_callback
+                    home_info.data["id"], self.on_message_callback
                 )
             async with async_timeout.timeout(10):
                 device_response = await self._client.get_auth_devices()
-                yl_devices: list[YoLinkDevice] = []
-                for device_info in device_response.data[ATTR_DEVICE]:
-                    yl_device = YoLinkDevice(device_info, self._client)
-                    yl_devices.append(yl_device)
-                self.yl_devices = yl_devices
+
         except YoLinkAuthFailError as yl_auth_err:
             raise ConfigEntryAuthFailed from yl_auth_err
+
         except (YoLinkClientError, asyncio.TimeoutError) as err:
             raise ConfigEntryNotReady from err
 
+        yl_devices: list[YoLinkDevice] = []
+
+        for device_info in device_response.data[ATTR_DEVICE]:
+            yl_devices.append(YoLinkDevice(device_info, self._client))
+
+        self.yl_devices = yl_devices
+
     async def fetch_device_state(self, device: YoLinkDevice):
         """Fetch Device State."""
-        async with async_timeout.timeout(20):
-            device_state_resp = await device.fetch_state_with_api()
-            if ATTR_DEVICE_STATE in device_state_resp.data:
-                self.data[device.device_id] = device_state_resp.data[ATTR_DEVICE_STATE]
-
-    async def _async_update_data(self) -> dict:
         try:
-            for yl_device in self.yl_devices:
-                await self.fetch_device_state(yl_device)
+            async with async_timeout.timeout(10):
+                device_state_resp = await device.fetch_state_with_api()
+                if ATTR_DEVICE_STATE in device_state_resp.data:
+                    self.data[device.device_id] = device_state_resp.data[
+                        ATTR_DEVICE_STATE
+                    ]
         except YoLinkAuthFailError as yl_auth_err:
             raise ConfigEntryAuthFailed from yl_auth_err
         except YoLinkClientError as yl_client_err:
-            # code 010203 for can't connect to device currently
+            # code 010203 for can't connect to device currently(ignore wait for device report state by mqtt)
             if yl_client_err.code != "010203":
                 raise UpdateFailed(
                     f"Error communicating with API: {yl_client_err}"
                 ) from yl_client_err
+
+    async def _async_update_data(self) -> dict:
+        fetch_tasks = []
+        for yl_device in self.yl_devices:
+            fetch_tasks.append(self.fetch_device_state(yl_device))
+        asyncio.gather(*fetch_tasks)
```

Comment:
Are we sure that there will always be at least one task here or should we add a check?

Reply:
Regarding the `async_timeout` that was removed, I remember now that the coordinator catches `asyncio.TimeoutError` by default during refresh so we can use `async_timeout` here if needed.

Reply:
i moved async_timeout to fetch_device_state  method

Reply:
We need to await gather too.
```suggestion
        if fetch_tasks:
            await asyncio.gather(*fetch_tasks)
```